streams/src/test/java/org/apache/kafka/streams/integration/AdjustStreamThreadCountTest.java

```diff
@@ -88,69 +89,96 @@ public void setup() {
+        addStreamStateChangeListener(kafkaStreams);
+        kafkaStreams.start();
+        waitForStateTransition(KafkaStreams.State.RUNNING);
+    }
+
     @Test
     public void shouldAddStreamThread() throws Exception {
-        try (final KafkaStreams kafkaStreams = new KafkaStreams(builder.build(), properties)) {
-            StreamsTestUtils.startKafkaStreamsAndWaitForRunningState(kafkaStreams);
-            final int oldThreadCount = kafkaStreams.localThreadsMetadata().size();
-            assertThat(kafkaStreams.localThreadsMetadata().stream().map(t -> t.threadName().split("-StreamThread-")[1]).sorted().toArray(), equalTo(new String[] {"1", "2"}));
-
-            final Optional<String> name = kafkaStreams.addStreamThread();
-
-            assertThat(name, not(Optional.empty()));
-            TestUtils.waitForCondition(
-                () -> kafkaStreams.localThreadsMetadata().stream().sequential()
-                        .map(ThreadMetadata::threadName).anyMatch(t -> t.equals(name.orElse(""))),
-                "Wait for the thread to be added"
-            );
-            assertThat(kafkaStreams.localThreadsMetadata().size(), equalTo(oldThreadCount + 1));
-            assertThat(
-                kafkaStreams
-                    .localThreadsMetadata()
-                    .stream()
-                    .map(t -> t.threadName().split("-StreamThread-")[1])
-                    .sorted().toArray(),
-                equalTo(new String[] {"1", "2", "3"})
-            );
-            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.REBALANCING, DEFAULT_DURATION);
-            waitForApplicationState(Collections.singletonList(kafkaStreams), KafkaStreams.State.RUNNING, DEFAULT_DURATION);
-        }
+        final int oldThreadCount = kafkaStreams.localThreadsMetadata().size();
+        assertThat(kafkaStreams.localThreadsMetadata().stream().map(t -> t.threadName().split("-StreamThread-")[1]).sorted().toArray(), equalTo(new String[] {"1", "2"}));
+
+        stateToTransitions.clear();
+        final Optional<String> name = kafkaStreams.addStreamThread();
+
+        assertThat(name, not(Optional.empty()));
+        TestUtils.waitForCondition(
+            () -> kafkaStreams.localThreadsMetadata().stream().sequential()
+                    .map(ThreadMetadata::threadName).anyMatch(t -> t.equals(name.orElse(""))),
+            "Wait for the thread to be added"
+        );
+        assertThat(kafkaStreams.localThreadsMetadata().size(), equalTo(oldThreadCount + 1));
+        assertThat(
+            kafkaStreams
+                .localThreadsMetadata()
+                .stream()
+                .map(t -> t.threadName().split("-StreamThread-")[1])
+                .sorted().toArray(),
+            equalTo(new String[] {"1", "2", "3"})
+        );
+
+        waitForStateTransition(KafkaStreams.State.REBALANCING);
+        waitForStateTransition(KafkaStreams.State.RUNNING);
```

Comment:
I think it would be better to wait until the Kafka Streams client id in state `RUNNING` and then verify if the history of the states transitions after adding the stream thread is first `REBALANCING` and then `RUNNING`. Currently, the order is not verified as far as I can see.

Reply:
Good suggestion! Added a `hasStateTransition` method to verify that. Thanks.